android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/KeyboardController.java

```diff
@@ -166,30 +171,62 @@ public void add(Keyboard newKeyboard) {
       return;
     }
 
-    for (int i=0; i<list.size(); i++) {
-      // Update existing keyboard entry
-      if (newKeyboard.equals(list.get(i))) {
-        Log.d(TAG, "Updating keyboard with newKeyboard");
-        list.set(i, newKeyboard);
-        return;
+    synchronized (list) {
+      for (int i = 0; i < list.size(); i++) {
+        // Update existing keyboard entry
+        if (newKeyboard.equals(list.get(i))) {
+          Log.d(TAG, "Updating keyboard with newKeyboard");
+          list.set(i, newKeyboard);
+          return;
+        }
       }
-    }
 
-    // Add new keyboard
-    list.add(newKeyboard);
+      // Add new keyboard
+      list.add(newKeyboard);
+    }
   }
 
+  /**
+   * Remove the keyboard at the specified index.
+   * KeyboardPickerActivity checks that index > 0
+   * But we'll allow OEM apps to remove any keyboard
+   * @param index
+   */
+  public void remove(int index) {
+    // Check initialized, and disallow removing default keyboard
```

Comment:
The comments here are a bit confusing, especially when compared to the code.  What allows the OEM apps to remove any keyboard here, but not Keyman itself?

Reply:
The Keyman app prevents users from removing the default sil_euro_latin keyboard so that we can always revert to it on KMW exceptions (avoiding the blank keyboard).

The FV app has a different requirement because it removes and rebuilds the entire "installed keyboards" list whenever the user updates keyboard selections in the app. And the FV app doesn't have a "default" app to revert to...

Reply:
Again, with a bit more emphasis:

> The comments **here** are a bit confusing, especially when compared to the code. What allows the OEM apps to remove any keyboard **here**, but not Keyman itself?

Reply:
I've simplified the description to `remove()`.

The view in KeyboardPickerActivity.java l.137 has the  comment about preventing the default keyboard from being removed.